project.php

```diff
@@ -478,9 +479,10 @@ function do_project_info_table()
             $sql = sprintf("
```

Comment:
Rather than fixing this SQL, let's replace it with:
```php
            $special_days = load_special_days();
            if(isset($special_days[$spec_code]))
            {
                $spec_display = $special_days[$spec_code]["display_name"];
            }
            else
            {
                $spec_display = "($spec_code)";
            }
```
from `pinc/special_colors.inc`.

Reply:
Rather fascinatingly, PHP 7.0's null coalescing is more powerful than I thought. It can do far more than just replacing tedious `if isset($foo) { $foo } else { $bar }`.

You can use it with chains of array derefs and it does what you'd hope it would do. Eg
```
  error_reporting(-1);
  $aa = [ "foo" => [
            "bar" => "baz",
            "boz" => ""
          ]
        ];
  var_dump(NULL["foo"]["bar"]??"other1"); # => "other1"
  var_dump($aa["foo"]["xar"]??"other2");  # => "other2"
  var_dump($aa["xoo"]["bar"]??"other3");  # => "other3"
  var_dump($aa["foo"]["bar"]??"other4");  # => "baz"
  var_dump($aa["foo"]["boz"]??"other5");  # => ""
  var_dump($x??"I didn't define x!");     # => "I didn't define x!"
```

And we can use it in this code too...

Reply:
Holy cow that's awesome! Let's use more of this!

Reply:
I think what's going on here is that PHP suppresses all errors, squashes them to null and terminates early whenever it evaluates the left hand side of a `??` (which is why the`$x??"Undefined x"` case works).

So in one sense it's a modern version of `@`, with all the pitfalls that go along with that. But yeah, I like it!